Source/Dafny/Dafny.atg

```diff
@@ -3821,17 +3839,51 @@ ParensExpression<out Expression e>
      var args = new List<ActualBinding>();
   .)
   "("                                        (. x = t; .)
-  [ ActualBindings<args> ]
+  [ TupleArgs<args> ]
   ")"
-  (. if (args.Count == 1) {
+  (. if (args.Count == 1 && !args[0].IsGhost) {
        if (args[0].FormalParameterName != null) {
          SemErr(args[0].FormalParameterName, "binding not allowed in parenthesized expression");
        }
        e = new ParensExpression(x, args[0].Actual);
      } else {
+       // Compute the actual position of ghost arguments
+       List<bool> argumentGhostness;
+       if (args.TrueForAll(arg => arg.FormalParameterName != null)) {
+         bool[] ghostness = new bool[args.Count];
+         bool[] initialized = new bool[args.Count];
+         for (int i = 0; i < initialized.Length; ++i) {
+           initialized[i] = false;
+         }
+         foreach (var arg in args) {
+           var success = int.TryParse(arg.FormalParameterName.val, out var index);
+           if (!success) {
+             SemErr(arg.FormalParameterName, "the name of tuple arguments must be an integer");
+           } else {
+             if (index < 0 || ghostness.Length <= index) {
+               SemErr(arg.FormalParameterName, "the name of tuple argument must describe a valid position in the tuple, starting from zero");
+             } else {
+               if (initialized[index]) {
+                 SemErr(arg.FormalParameterName, "different tuple arguments cannot have the same name");
+               } else {
+                 ghostness[index] = arg.IsGhost;
+                 initialized[index] = true;
```

Comment:
This is too specific. The code here only needs to find the right tuple types _for correctly given tuples_. If parameters are duplicated or misspelled, the Resolver will complain later. I suggest the following logic:

```
var ghostness = new bool[args.Count];
for (var i = 0; i < args.Count; i++) {
  ghostness[i] = true;
}
for (var i = 0; i < args.Count; i++) {
  var arg = args[i];
  if (arg.IsGhost) {
    if (arg.FormalParameterName == null) {
      ghostness[i] = true;
    } else {
      var success = int.TryParse(arg.FormalParameterName.val, out var index);
      if (success && 0 <= index && index < args.Count) {
        ghostness[index] = true;
      }
    }
  }
}
var argumentGhostness = ghostness.ToList();
```

Reply:
Note, btw, that a parameter with unnecessary leading zeros (e.g., `007`) is incorrect. `TryParse` will succeed, but that's okay, because the Resolver will find the error later.

Reply:
Changed, thanks.